### PR TITLE
add existing group to group

### PR DESCRIPTION
**Pull request by elduddz**
- if a group exists it will be added as a group in the UG / DL (depending on primary)

- Groups that are created in the scope of the same ADM run are not available to use.

### CODE REVIEW

File: UKHO.ADM/internal/ConvertTo-GroupObjects.ps1 [hunk 1]
```diff
@@ -28,16 +38,27 @@ function ConvertTo-GroupObjects {
...
+                }
+
                 $groups += $ugGroup
                 $groups += $ggGroup
             }
 
             #Add users
-            if ($GroupConfig.Users -ne $null -and $GroupConfig.Users.Count -gt 0) { 
-                $GroupConfig.Users | Where {$_.Domain -eq $OU.Domain.FQDN} | ForEach-Object {
+            if ($null -ne $GroupConfig.Users -and $GroupConfig.Users.Count -gt 0) {
```
Comment (MarcDenman) [inline UKHO.ADM/internal/ConvertTo-GroupObjects.ps1:42]: Why the inversion for `$var -ne $null` to `$null -ne $var`?

Comment (elduddz) [reply UKHO.ADM/internal/ConvertTo-GroupObjects.ps1:42]: PSScriptAnalyser suggests that as a problem;
$null should be on the left side of equality comparisons.

Comment (elduddz) [reply UKHO.ADM/internal/ConvertTo-GroupObjects.ps1:42]: https://rencore.com/blog/powershell-null-comparison/

tl;dr
Use $null in comparison on the left hand side. Due to type inference Arrays will be checked for $null with each elements total null count equating.

File: UKHO.ADM/internal/ConvertTo-GroupObjects.ps1 [hunk 2]
```diff
@@ -20,6 +20,16 @@ function ConvertTo-GroupObjects {
                 $ggGroup = [ADGroup]::new($Settings.GroupPrefix, $Settings.Environment, $OU.Name, $GroupConfig.Name, "GG", $OU.DistinguishedName, $OU.Domain)
                 $dlGroup.ADGroupMembers += $ggGroup
 
+                # Add Other Groups
+                if ($null -ne $GroupConfig.Groups -and $GroupConfig.Groups.Count -gt 0) {
+                    $GroupConfig.Groups | ForEach-Object {
+                        $groupGroup = [ADGroup]::new($_.DistinguishedName, $OU.Domain)
+                        if (Check-GroupExists $groupGroup) {
+                            $dlGroup.ADGroupMembers += $groupGroup
```
Comment (MarcDenman) [inline UKHO.ADM/internal/ConvertTo-GroupObjects.ps1]: From the users perspective, I don't know if this "Check-GroupExists" is useful. If the user has made a typo this will silently not add the group which could be very annoying as it appear that ADM is just ignoring that group.

If we do check for the groups existence here, can we also log that the group doesn't exist so won't be added? 

The similar scenario is `Check-UserExists` which is invoked in [AddUserToG](https://github.com/UKHO/ADM-PSmodule/blob/328e6d726700409a66390cb6b7de1f6f9c5522bd/UKHO.ADM/Objects/Objects.ps1#L203) so you might be able to copy that existing pattern. But that would only throw the error at "invoke" time which would be annoying.

File: UKHO.ADM/internal/ConvertTo-GroupObjects.ps1 [hunk 3]
```diff
@@ -28,16 +38,27 @@ function ConvertTo-GroupObjects {
                 $ggGroup = [ADGroup]::new($Settings.GroupPrefix, $Settings.Environment, $OU.Name, $GroupConfig.Name, "GG", $OU.DistinguishedName, $OU.Domain)
                 $ugGroup.ADGroupMembers += $ggGroup
 
+                # Add Other Groups
+                if ($null -ne $GroupConfig.Groups -and $GroupConfig.Groups.Count -gt 0) {
+                    $GroupConfig.Groups | ForEach-Object {
+                        $groupGroup = [ADGroup]::new($_.DistinguishedName, $_.Domain)
+                        if (Check-GroupExists $groupGroup) {
```
Comment (MarcDenman) [inline UKHO.ADM/internal/ConvertTo-GroupObjects.ps1]: This `Check-GroupExists` faces the same question as above.

File: UKHO.ADM/internal/ConvertTo-GroupObjects.ps1 [hunk 4]
```diff
@@ -28,16 +38,27 @@ function ConvertTo-GroupObjects {
...
+                            $ugGroup.ADGroupMembers += $groupGroup
```
Comment (MarcDenman) [inline UKHO.ADM/internal/ConvertTo-GroupObjects.ps1]: Should this add to the GG instead of the UG? I am not 100% either way. Looking at our AD on a non primary domain, we add the users to the GG, this is then added to the UG. Does it make more sense for the group to be added to GG so access is "defined in the same place"? Might be worth double checking with @thealmguy

File: UKHO.ADM/internal/ConvertTo-GroupObjects.ps1 (hunk 2 above)
Comment (MarcDenman) [inline UKHO.ADM/internal/ConvertTo-GroupObjects.ps1]: Same question as above about whether this adds to the correct group. Should this be the GG instead of the DL?

**Review by MarcDenman — CHANGES_REQUESTED** [review]
Does this assume that the group exists in each domain? Is that true? Would it make sense to be able to specify the domain a group exists in from the configuration?

I am impressed with how little work this is to add in this functionality actually.

Can you please add some documentation around this? It would be nice to have an example which includes this to help show the expected format in the configuration

If possible, please add a few tests.

File: UKHO.ADM/internal/ConvertTo-GroupObjects.ps1 (hunk 1 above)
Comment (MarcDenman) [reply UKHO.ADM/internal/ConvertTo-GroupObjects.ps1:42]: Huh. OK. Fair enough, if that is what is PSScripAnalyser suggests.

File: UKHO.ADM/internal/ConvertTo-GroupObjects.ps1 (hunk 4 above)
Comment (elduddz) [reply UKHO.ADM/internal/ConvertTo-GroupObjects.ps1]: probably should go in the GG, as if you added the GG to somewhere, the group would be missed

File: UKHO.ADM/internal/ConvertTo-GroupObjects.ps1 (hunk 2 above)
Comment (elduddz) [reply UKHO.ADM/internal/ConvertTo-GroupObjects.ps1]: this makes it the same code, which is even more amazing

File: UKHO.ADM/internal/ConvertTo-GroupObjects.ps1 (hunk 3 above)
Comment (elduddz) [reply UKHO.ADM/internal/ConvertTo-GroupObjects.ps1]: until you get to the right domain the group doesnt exist...

**Comment by elduddz** [comment]
missed an extra Object builder off, will get this working local first :)

File: UKHO.ADM/internal/ConvertTo-GroupObjects.ps1 (hunk 2 above)
Comment (elduddz) [reply UKHO.ADM/internal/ConvertTo-GroupObjects.ps1]: doing it, written to Write-Verbose pipeline

File: UKHO.ADM/internal/ConvertTo-GroupObjects.ps1 (hunk 4 above)
Comment (elduddz) [reply UKHO.ADM/internal/ConvertTo-GroupObjects.ps1]: added to GG level

File: UKHO.ADM/internal/ConvertTo-GroupObjects.ps1 (hunk 2 above)
Comment (elduddz) [reply UKHO.ADM/internal/ConvertTo-GroupObjects.ps1]: agreed :100:

**Comment by elduddz** [comment]
ready for another review when you are ready ;)

**Review by MarcDenman — APPROVED** [review]
